src/TextInputWidget.cc

```diff
@@ -101,6 +105,39 @@ FilteredTextEdit::keyPressEvent(QKeyEvent *event)
         }
 }
 
+bool
+FilteredTextEdit::canInsertFromMimeData(const QMimeData *source) const
+{
+        return (source->hasImage() || source->hasText() ||
+                QTextEdit::canInsertFromMimeData(source));
+}
+
+void
+FilteredTextEdit::insertFromMimeData(const QMimeData *source)
+{
+        if (source->hasImage()) {
+                QPixmap const img = qvariant_cast<QPixmap>(source->imageData());
+                previewDialog_    = new dialogs::PreviewImageOverlay{img, this};
+                connect(previewDialog_,
+                        &dialogs::PreviewImageOverlay::confirmImageUpload,
+                        this,
+                        &FilteredTextEdit::receiveImage);
+                previewDialog_->show();
+        } else if (source->hasFormat("x-special/gnome-copied-files") &&
+                   QImageReader{source->text()}.canRead()) {
```

Comment:
What if a user pastes in non-image files?

Reply:
As of now, pasting non-image files will trigger the else clause and therefore, Qt's default behaviour will take over. I think adding support for pasting non-image files (as well as audio, video, etc) is fairly simple, and probably out of the scope of this PR. It is definitely something I can look into in the future.